src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -0,0 +1,280 @@
+
+                var codeDocument = projectEngine.Process(projectItem);
+                var csharpDocument = codeDocument.GetCSharpDocument();
+                for (var j = 0; j < csharpDocument.Diagnostics.Count; j++)
+                {
+                    var razorDiagnostic = csharpDocument.Diagnostics[j];
+                    var csharpDiagnostic = razorDiagnostic.AsDiagnostic();
+                    context.ReportDiagnostic(csharpDiagnostic);
+                }
+
+                var hint = GetIdentifierFromPath(file.NormalizedPath);
+
+                var generatedCode = csharpDocument.GeneratedCode;
+                if (razorContext.WriteGeneratedContent)
+                {
+                    var path = file.GeneratedOutputPath;
+                    Directory.CreateDirectory(Path.GetDirectoryName(path));
+                    File.WriteAllText(path, generatedCode);
+                }
+
+                outputs[i] = (hint, SourceText.From(generatedCode, Encoding.UTF8));
+            });
+
+            for (var i = 0; i < files.Count; i++)
+            {
+                var (hint, sourceText) = outputs[i];
+                context.AddSource(hint, sourceText);
+            }
+
+            arraypool.Return(outputs);
+        }
+
+        private static IReadOnlyList<TagHelperDescriptor> ResolveTagHelperDescriptors(GeneratorExecutionContext GeneratorExecutionContext, RazorSourceGenerationContext razorContext)
+        {
+            var tagHelperFeature = new StaticCompilationTagHelperFeature();
+
+            var langVersion = ((CSharpParseOptions)GeneratorExecutionContext.ParseOptions).LanguageVersion;
+
+            var discoveryProjectEngine = RazorProjectEngine.Create(razorContext.Configuration, razorContext.FileSystem, b =>
+            {
+                b.Features.Add(new DefaultTypeNameFeature());
+                b.Features.Add(new ConfigureRazorCodeGenerationOptions(options =>
+                {
+                    options.SuppressPrimaryMethodBody = true;
+                    options.SuppressChecksum = true;
+                }));
+
+                b.SetRootNamespace(razorContext.RootNamespace);
+
+                var metadataReferences = new List<MetadataReference>(GeneratorExecutionContext.Compilation.References);
+                b.Features.Add(new DefaultMetadataReferenceFeature { References = metadataReferences });
+
+                b.Features.Add(tagHelperFeature);
+                b.Features.Add(new DefaultTagHelperDescriptorProvider());
+
+                CompilerFeatures.Register(b);
+                RazorExtensions.Register(b);
+
+                b.SetCSharpLanguageVersion(langVersion);
+            });
+
+            var files = razorContext.RazorFiles;
+            var results = ArrayPool<SyntaxTree>.Shared.Rent(files.Count);
+
+            var parseOptions = (CSharpParseOptions)GeneratorExecutionContext.ParseOptions;
+
+            Parallel.For(0, files.Count, GetParallelOptions(), i =>
```

Comment:
Please make sure to pass down `context.CancellationToken` for parallelization work like this.